hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/commit/BaseWriteHelper.java

```diff
@@ -85,9 +99,58 @@ public I combineOnCondition(
-    return deduplicateRecords(records, table.getIndex(), parallelism, table.getConfig().getSchema(), table.getConfig().getProps(), recordMerger);
+    RecordMergeMode recordMergeMode = HoodieTableConfig.inferCorrectMergingBehavior(null, table.getConfig().getPayloadClass(), null,
+        String.join(",", orderingFieldNames), tableConfig.getTableVersion()).getLeft();
+    BufferedRecordMerger<T> bufferedRecordMerger = BufferedRecordMergerFactory.create(
+        readerContext,
+        recordMergeMode,
+        false,
+        Option.ofNullable(recordMerger),
+        orderingFieldNames,
+        Option.ofNullable(table.getConfig().getPayloadClass()),
+        new SerializableSchema(table.getConfig().getSchema()).get(),
+        table.getConfig().getProps(),
+        tableConfig.getPartialUpdateMode());
+    return deduplicateRecords(
+        records,
+        table.getIndex(),
+        parallelism,
+        table.getConfig().getSchema(),
+        table.getConfig().getProps(),
+        bufferedRecordMerger,
+        readerContext,
+        orderingFieldNames.toArray(new String[0]));
   }
 
-  public abstract I deduplicateRecords(I records, HoodieIndex<?, ?> index, int parallelism, String schema, TypedProperties props, HoodieRecordMerger merger);
+  public abstract I deduplicateRecords(I records,
+                                       HoodieIndex<?, ?> index,
+                                       int parallelism,
+                                       String schema,
+                                       TypedProperties props,
+                                       BufferedRecordMerger<T> merger,
+                                       HoodieReaderContext<T> readerContext,
+                                       String[] orderingFieldNames);
+
+  public static <T> Option<BufferedRecord<T>> merge(HoodieRecord<T> newRecord,
+                                                    HoodieRecord<T> oldRecord,
+                                                    Schema newSchema,
+                                                    Schema oldSchema,
+                                                    RecordContext<T> recordContext,
+                                                    String[] orderingFieldNames,
+                                                    BufferedRecordMerger<T> recordMerger,
+                                                    TypedProperties properties) throws IOException {
+    // Construct new buffered record.
+    BufferedRecord<T> bufferedRec1 = BufferedRecord.forRecordWithContext(newRecord, newSchema, recordContext, properties, orderingFieldNames);
+    // Construct old buffered record.
+    BufferedRecord<T> bufferedRec2 = BufferedRecord.forRecordWithContext(oldRecord, oldSchema, recordContext, properties, orderingFieldNames);
+    // Run merge.
+    return recordMerger.deltaMerge(bufferedRec1, bufferedRec2);
```

Comment:
did we not discuss we should retain the precombine semantics here. 
and hence we can't call into deltaMerge directly, but just poll for ordering field and choose rec1 or rec2?